Issue#1844

x7FFFFFFF opened the pull request:
https://github.com/h2database/h2database/issues/1844
Mysql Compatibility: create table error when primary key has comment

paramonov-alex committed:
Issue#1844
https://github.com/h2database/h2database/issues/1844
Mysql Compatibility: create table error when primary key has comment

katzyn commented on h2/src/test/org/h2/test/scripts/ddl/createTable.sql:
```diff
@@ -131,3 +131,9 @@ SELECT COLUMN_NAME, COLUMN_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME
 DROP TABLE TEST;
 > ok
+
+CREATE TABLE IF NOT EXISTS TEST(ID BIGINT(20) NOT NULL AUTO_INCREMENT PRIMARY KEY COMMENT '字典ID', FIELD_NAME VARCHAR(100) NOT NULL COMMENT '字段名称', TABLE_NAME VARCHAR(100) NOT NULL COMMENT '表名');
```
Please, do not use non-ASCII characters in test cases, they aren't needed here.

Also it's reasonable to test that comment was really set with `SCRIPT TABLE TEST` command or in some other way.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -8029,6 +8029,10 @@ private void parseTableColumnDefinition(CommandWithColumns command, Schema schem
                 parseReferences(ref, schema, tableName);
                 command.addConstraintCommand(ref);
             }
+            String comment = readCommentIf();
```
It looks like comment should be parsed **before** `CHECK` and `REFERENCES`.
https://dev.mysql.com/doc/refman/8.0/en/create-table.html
```SQL
column_definition:
    data_type [NOT NULL | NULL] [DEFAULT {literal | (expr)} ]
      [AUTO_INCREMENT] [UNIQUE [KEY]] [[PRIMARY] KEY]
      [COMMENT 'string']
      [COLLATE collation_name]
      [COLUMN_FORMAT {FIXED|DYNAMIC|DEFAULT}]
      [STORAGE {DISK|MEMORY}]
      [reference_definition]
      [check_constraint_definition]
  | data_type
      [COLLATE collation_name]
      [GENERATED ALWAYS] AS (expr)
      [VIRTUAL | STORED] [NOT NULL | NULL]
      [UNIQUE [KEY]] [[PRIMARY] KEY]
      [COMMENT 'string']
      [reference_definition]
      [check_constraint_definition]
```

Also comment can be already read by `parseColumnForTable()`. In this situation it should not be read second time to forbid definitions like `columnName INT COMMENT 'some comment 1' COMMENT 'some comment 2'`.

paramonov-alex committed:
Issue#1844
https://github.com/h2database/h2database/issues/1844
https://github.com/h2database/h2database/pull/1877#discussion_r277528585
https://github.com/h2database/h2database/pull/1877#discussion_r277531220

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -8013,6 +8013,15 @@ private void parseTableColumnDefinition(CommandWithColumns command, Schema schem
             }
+            String comment = readCommentIf();
+            if (comment != null) {
+                if (column.getComment() != null) {
```
These two conditions should be simply swapped.

`readCommentIf()` should be invoked only when comment was not already set.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -8013,6 +8013,15 @@ private void parseTableColumnDefinition(CommandWithColumns command, Schema schem
+                if (column.getComment() != null) {
+                    addExpected(CHECK);
+                    addExpected("REFERENCES");
+                    throw getSyntaxError();
```
These three lines throw incomplete exception for duplicate comment.

But with swapped conditions they should be simply removed.

katzyn commented on h2/src/test/org/h2/test/db/TestRunscript.java:
```diff
@@ -51,6 +51,7 @@ public void test() throws Exception {
         testCancelScript();
         testEncoding();
         testClobPrimaryKey();
+        testComment();
```
Please, move this test to `createTable.sql`. `SCRIPT TABLE TEST1` output is not too big and I don't see any reason to use custom checks for presence of substrings.

x7FFFFFFF replied:
How to skip "CREATE USER " row comparison in createTable.sql ?
 Every time i run test salt and hash are different.
Example:

```
SCRIPT TABLE TEST1;
> SCRIPT
> ----------------------------------------------------------------------------------------------------------------------
> -- 0 +/- SELECT COUNT(*) FROM PUBLIC.TEST1;
> ALTER TABLE "PUBLIC"."TEST1" ADD CONSTRAINT "PUBLIC"."CONSTRAINT_4" PRIMARY KEY("ID");
> CREATE CACHED TABLE "PUBLIC"."TEST1"( "ID" BIGINT NOT NULL COMMENT 'COMMENT1', "FIELD_NAME" VARCHAR(100) NOT NULL COMMENT 'COMMENT2' );
> CREATE USER IF NOT EXISTS "SA" SALT '7f0c86261a0b7222' HASH '920138111dd8344f1d4520e6f99a58040e852b7361d552397f1cd4f18a6debc1' ADMIN;
> rows: 4
```

katzyn replied:
Try
```SQL
SCRIPT NODATA NOPASSWORDS NOSETTINGS TABLE TEST
```
It should create reproducible results.

katzyn replied:
`NODATA` is not really required, of course, because your table is empty.

paramonov-alex committed:
Issue#1844
https://github.com/h2database/h2database/issues/1844
https://github.com/h2database/h2database/pull/1877#discussion_r277557709
https://github.com/h2database/h2database/pull/1877#discussion_r277559247
https://github.com/h2database/h2database/pull/1877#discussion_r277560394

paramonov-alex committed:
Issue#1844
https://github.com/h2database/h2database/issues/1844
revert to e9cafd8140bc2cfbe7ba040175c34efbe6356834

katzyn commented on h2/src/test/org/h2/test/scripts/ddl/createTable.sql:
```diff
@@ -131,3 +131,27 @@ SELECT COLUMN_NAME, COLUMN_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME
 DROP TABLE TEST;
 > ok
+
+CREATE TABLE TEST1(ID BIGINT(20) NOT NULL PRIMARY KEY COMMENT 'COMMENT1', FIELD_NAME VARCHAR(100) NOT NULL COMMENT 'COMMENT2');
```
Use `CREATE MEMORY TABLE` here, because by default it can be either `CACHED` or `MEMORY` that is not expected by the following `SCRIPT` test.

katzyn commented on h2/src/test/org/h2/test/scripts/ddl/createTable.sql:
```diff
@@ -131,3 +131,27 @@ SELECT COLUMN_NAME, COLUMN_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME
+> ---------------------------------------------------------------------------------------------------------------------------------------
+> -- 0 +/- SELECT COUNT(*) FROM PUBLIC.TEST1;
+> ALTER TABLE "PUBLIC"."TEST1" ADD CONSTRAINT "PUBLIC"."CONSTRAINT_4" PRIMARY KEY("ID");
+> CREATE CACHED TABLE "PUBLIC"."TEST1"( "ID" BIGINT NOT NULL COMMENT 'COMMENT1', "FIELD_NAME" VARCHAR(100) NOT NULL COMMENT 'COMMENT2' );
```
And adjust expected output too.

katzyn reviewed (commented):
Please, send a license statement as described here:
https://h2database.com/html/build.html#providing_patches
It's better to send it to our mailing list:
https://groups.google.com/forum/#!forum/h2-database
Mailing list is partially pre-moderated, your message may not appear immediately.

There is a one more minor problem that needs to be fixed. See comments below.

paramonov-alex committed:
Issue#1844
https://github.com/h2database/h2database/issues/1844
https://github.com/h2database/h2database/pull/1877#discussion_r277630517

katzyn commented:
Thank you!
